platforms/posix/src/px4/common/main.cpp

```diff
@@ -572,29 +575,33 @@ void print_usage()
 	printf("        e.g.: px4-commander status\n");
 }
 
-bool is_already_running(int instance)
+bool is_already_running(int instance, bool keep_locked)
 {
 	const std::string file_lock_path = std::string(LOCK_FILE_PATH) + '-' + std::to_string(instance);
-	struct flock fl;
 	int fd = open(file_lock_path.c_str(), O_RDWR | O_CREAT, 0666);
 
 	if (fd < 0) {
+		PX4_ERR("is_already_running: failed to create lock file: %s, reason=%s", file_lock_path.c_str(), strerror(errno));
 		return false;
 	}
 
-	fl.l_type   = F_WRLCK;
-	fl.l_whence = SEEK_SET;
-	fl.l_start  = 0;
-	fl.l_len    = 0;
-	fl.l_pid    = getpid();
+	PX4_DEBUG("is_already_running: process %d is grabbing lock...", getpid());
+	if (flock(fd, LOCK_EX) < 0) {
+		// if we cannot get the exclusive lock, then it is already running!
+		PX4_ERR("is_already_running: failed to get lock on file: %s, reason=%s", file_lock_path.c_str(), strerror(errno));
+		return true;
+	}
+	PX4_DEBUG("is_already_running: file locked !");
 
-	if (fcntl(fd, F_SETLK, &fl) == -1) {
-		// We failed to create a file lock, must be already locked.
-		return errno == EACCES || errno == EAGAIN;
+	if (!keep_locked) {
+		flock(fd, LOCK_UN);
+		PX4_DEBUG("is_already_running: releasing the lock !");
+		errno = 0;
+		return false;
 	}
 
 	errno = 0;
-	return false;
+	return true;
```

Comment:
How is this supposed to work? The server releases the lock, and clients get here and just assume the server is running?

Reply:
Fixed, I added a flag to tell the helper function whether this is the client or server calling, so the client can use a non-blocking lock operation to tell if the server is running.